### ISSUE TITLE
NK3 not detected/discovered at OEM Factory Reset/Re-Ownership?

### ISSUE THREAD

**Issue by tlaurion** [Nitrokey/heads#48]
Edit: fixed with https://github.com/linuxboot/heads/commit/67f1dae840446b32d323248c614ac84ee90c880f

Cause: race condition between USB controller kernel modules being loaded and the USB Security dongle being discovered and used. By moving the sleep function after all USB controller kernel modules are loaded, the issue disappears.
 
----------

~Sorry for lack of details here. It was reported that Heads was not able to properly do OEM-Factory reset/Re-Ownership on nv41 because the dongle (nk3) was first detected (those checks have been moved earlier in oem-factory-reset for dongle type detection to choose algo) but when comes the time to factory reset the dongle and generate keys inside of them, it fails.~

~I initially thought physical presence was the issue (touch the key) but tests showed that the dongle was then properly used while moving the dongle around on different usb ports (on a x230).~

~I'm wondering what happened in the NK3 firmware maybe?~ Just slower to be brought up
~Searched and didn't find any issue opened. Any hint on what is happening? Is it possibly linked to firmware version that was flashed on the NK3? How users are supposed to know, what are the guidelines for fixing this? Are users supposed to know how to flash NK3 firmware? This behavior was not present on NK2 pro/NK2/NK1 (which I stopped using after moving my key backup to nk3 after non-default USB thumb drive key material backup that landed under master a while ago).~

TLDR: random issues with NK3 at Re-Ownership vs OEM factory reset from OEM pre-shipment?
~A work group tackling those issues would be needed, since its NK3 related (firmware cannot do much here. If NK3 dongle detected early in oem-factory-reset, Heads expects to be able to talk to the dongle when factory resetting it through gpg2 calls and then be able to change PINs. That seems to be where the problem is today, which might have missed proper QA since no re-ownership, or provisioning with non-default PINs, which requires talking with the dongle for PIN changes?)~

~@jans23 @szszszsz @daringer  ?~
~Can you do some testing on your side? Nothing explains behavior change on Heads side of things, so needs to be a NK3 related firmware thing? Not sure how to troubleshoot isolate the problem here for resolution.~

**Comment by tlaurion** [comment]
Also, hotp-verification hash upstream points to later available commit while nitrokey fork points to 1.4?

Upstream:
https://github.com/linuxboot/heads/blob/ee1978ffc0afa4db1b0a46517aff946912cdca14/modules/hotp-verification#L5-L10
 
Nitrokey:
https://github.com/Nitrokey/heads/blob/da154df82c61f5b3ff2baad540f55500da2373d1/modules/hotp-verification#L5-L10

Was the latest hotp-verification code not aimed to be used upstream @daringer ?

**Comment by daringer** [comment]
That's weird, we don't see any issues currently - but let's do some additional testing and/or verification ... 
can you tell if the nk3 is properly starting in the not-working-scenario:
* you should see that the nk3 LED goes white shortly after being powered-on and afterwards the expectation is it should stay off until an action happens
* do you observe any LED activity during the not working oem-factory reset?

**Comment by tlaurion** [comment]
> That's weird, we don't see any issues currently - but let's do some additional testing and/or verification ... can you tell if the nk3 is properly starting in the not-working-scenario:
> 
> * you should see that the nk3 LED goes white shortly after being powered-on and afterwards the expectation is it should stay off until an action happens

Yes. nk3a NFC v.1.5.0 (can;t upgrade, will open new issue)

> * do you observe any LED activity during the not working oem-factory reset?

No led activity. Nothing happens until I move the NK3 to another port and restart the wizard.

> * this might also boil down to a hardware issue, maybe even with x230 ? do you observe consistent failure on a specific port, I remember we had something like this before reported by some customer

Another customer is tempting to troubleshoot the issue for a little while under https://matrix.to/#/!RNcjJXCGHiyxXCHpKv:matrix.org/$20rR2YYEZlMs9iD1w9giODaCzqtz1ERSz7iBA8C7gLk?via=matrix.org&via=nitro.chat&via=invisiblethingslab.com

And I could replicate.
> 
> @alexgithublab can you please try to reproduce the reported behavior with our firmware and compare against an upstream build from here: [nv41](https://app.circleci.com/pipelines/github/linuxboot/heads/750/workflows/c1cdc18d-c537-4040-9f6d-f92b2cc35cae/jobs/15812/artifacts) or [ns50](https://app.circleci.com/pipelines/github/linuxboot/heads/750/workflows/c1cdc18d-c537-4040-9f6d-f92b2cc35cae/jobs/15832/artifacts) or [x230](https://app.circleci.com/pipelines/github/linuxboot/heads/750/workflows/c1cdc18d-c537-4040-9f6d-f92b2cc35cae/jobs/15801/artifacts)

**Comment by tlaurion** [comment]
Seems like https://github.com/tlaurion/heads/commit/78fe0f9ab3fd126a8128ca0f6a291f65809e4b06

Fixes it. Weird that the timing was not an issue before and become one.

**Comment by alexgithublab** [comment]
OEM factory reset works with this [image](https://app.circleci.com/pipelines/github/linuxboot/heads/750/workflows/c1cdc18d-c537-4040-9f6d-f92b2cc35cae/jobs/15812/artifacts) (with the sleep commit) on the x230 and NK3 1.6.0

**Comment by tlaurion** [comment]
@alexgithublab please tag me or refer to PR or better, comment upstream telling that à PR fix an issue to steamline review and fasten PR merges.

**Comment by tlaurion** [comment]
Seems like the real issue is https://github.com/Nitrokey/nitrokey-pro-firmware/issues/54 and/or https://github.com/Nitrokey/libnitrokey/issues/137

**Comment by tlaurion** [comment]
> No doubt https://github.com/linuxboot/heads/pull/1638 helped, but once gpg --card-status calls happen followed by accesses to libnitrokey library usage, the problem reappears.
> 
> Seems like the real issue is https://github.com/Nitrokey/nitrokey-pro-firmware/issues/54 and underlying https://github.com/Nitrokey/libnitrokey/issues/137
> 
> @sosthene-nitrokey @szszszsz ?

From investigation on nk3 alone, it doesn't seem to be true.

Unfortunately without further work from Nitrokey, results from hotp_verification are invalid as of now, and it's output is taken for decision making under Heads. Pleas cinfsttme off-channel, I already spent too much time into this.

**Comment by tlaurion** [comment]
@daringer : this was fixed upsteam 3 weeks ago. Modified OP https://github.com/Nitrokey/heads/issues/48#issue-2234019575

Customers are expecting this to be fixed in forks
Cross-ref matrix discussion: https://matrix.to/#/!RNcjJXCGHiyxXCHpKv:matrix.org/$9IMiPjkK-XujuiZ2yMlv7Fr1fkvdlQaoXIcMkCm7KHo?via=matrix.org&via=nitro.chat&via=invisiblethingslab.com

**Comment by wessel-novacustom** [comment]
> @daringer @wessel-novacustom: this was fixed upsteam 3 weeks ago. Modified OP [#48 (comment)](https://github.com/Nitrokey/heads/issues/48#issue-2234019575)
> 
> Customers are expecting this to be fixed in forks Cross-ref matrix discussions (require to be Dasharo subscriber to access):
> 
>     * https://matrix.to/#/!RNcjJXCGHiyxXCHpKv:matrix.org/$9IMiPjkK-XujuiZ2yMlv7Fr1fkvdlQaoXIcMkCm7KHo?via=matrix.org&via=nitro.chat&via=invisiblethingslab.com
> 
>     * https://matrix.to/#/!RNcjJXCGHiyxXCHpKv:matrix.org/$uuc-UIYgByDH5sImAK3DxYQdBjOrjZz1WePQtV9izZE?via=matrix.org&via=nitro.chat&via=invisiblethingslab.com

Thank you for your work!

@macpijan We need to make sure that important fixes and improvements like these are being added to the next release.
